src/crimson/net/SocketConnection.cc

```diff
@@ -197,31 +183,26 @@ seastar::future<MessageRef> SocketConnection::do_read_message()
       // read footer
       return socket->read(sizeof(m.footer));
     }).then([this] (bufferlist bl) {
-      // resume background processing of tags
-      read_tags_until_next_message();
-
       auto p = bl.cbegin();
       ::decode(m.footer, p);
       auto msg = ::decode_message(nullptr, 0, m.header, m.footer,
                                   m.front, m.middle, m.data, nullptr);
       // TODO: set time stamps
       msg->set_byte_throttler(policy.throttler_bytes);
-      constexpr bool add_ref = false; // Message starts with 1 ref
-      return MessageRef{msg, add_ref};
-    });
-}
 
-seastar::future<MessageRef> SocketConnection::read_message()
-{
-  return seastar::repeat_until_value([this] {
-      return do_read_message()
-        .then([this] (MessageRef msg) -> std::optional<MessageRef> {
-          if (!update_rx_seq(msg->get_seq())) {
-            // skip this request and read the next
-            return {};
-          }
-          return msg;
+      if (!update_rx_seq(msg->get_seq())) {
+        // skip this message
+        return;
+      }
+
+      constexpr bool add_ref = false; // Message starts with 1 ref
+      auto msg_ref = MessageRef{msg, add_ref};
+      // start dispatch, ignoring exceptions from the application layer
+      seastar::with_gate(pending_dispatch, [this, msg = std::move(msg_ref)] {
+          return dispatcher.ms_dispatch(this, std::move(msg))
+            .handle_exception([] (std::exception_ptr eptr) {});
```

Comment:
i agree that we're not responsible for handling the exceptions here. but i don't think we should not ignore it either. how about dump the the exception and `assert()`? so the (buggy) application developer could understand what's going wrong?

Reply:
Agree, added error logging and `ceph_assert()`.